Retry when secret is referred by machineClass is missing

AxiomSamarth opened the pull request:
**What this PR does / why we need it**:
The PR returns an `error` when the Secret referred by the MachineClass is absent/not created yet. Waits until the secret is created while prompting it on the logs. MachineDeployment immediately reconciles when the referred secret is available later.

**Which issue(s) this PR fixes**:
Fixes #430 

**Special notes for your reviewer**:
This is my first PR to this repo. I have tried my best to adhere to the guidelines. Let me know if there is anything to be rectified.

**Release note**:
<!--  Write your release note:
1. Enter your release note in the below block.
2. If no release note is required, just write "NONE" within the block.

Format of block header: <category> <target_group>
Possible values:
- category:       improvement|noteworthy
- target_group:   user|operator|developer
-->
```improvement user
Retry when secret is referred by machineClass is missing
```
 @hardikdr @prashanth26

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/gardener/machine-controller-manager?pullRequest=495) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/gardener/machine-controller-manager?pullRequest=495) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/gardener/machine-controller-manager?pullRequest=495) it.</sub>

gardener-robot-ci-1 commented:
Thank you @AxiomSamarth for your contribution. Before I can start building your PR, a member of the organization must set the required label(s) {'reviewed/ok-to-test'}. Once started, you can check the build status in the PR checks section below.

ggaurav10 commented on pkg/controller/controller.go:
```diff
@@ -548,6 +548,10 @@ func worker(queue workqueue.RateLimitingInterface, resourceType string, maxRetri
 					return false
 				}
 
+				for err != nil {
+					err = reconciler(key.(string))
+				}
+
```
why is this required?

AxiomSamarth replied:
Hi Gaurav, this loop is to wait until the reconcile error is resolved and the secret is created before proceeding. Seeing your comment, I realise that this loop will run even in the error scenario due to absence of machine class (not found or deleted).

As per my discussions with @hardikdr this is a temporary fix. A better and a more systematic fix with a retry limit is to be done in the near future.

hardikdr replied:
This is certainly a stopgap solution, we would eventually need a mechanism to exponentially-backoff for all the failing operations. 

Though I would suggest to not modify the worker-logic here, it could potentially block the worker-thread in the cases of errors.  

With this solution, we should simply trigger the reconciliation if the secret is NotFound, and that should be it. :)

AxiomSamarth replied:
Okay, I will update it likewise.

AxiomSamarth replied:
I have pushed my latest changes to reconcile it one more time if the secret referred by `machineClass` is `NotFound` instead of a wait loop.

prashanth26 replied:
> Though I would suggest to not modify the worker-logic here, it could potentially block the worker-thread in the cases of errors.

I would second with Hardik on this. I don't think we should modify this part of the logic. Because this would affect reconcialtion of all workers. i.e - Machines, MachineSet, MachineDeployments, MachinesClasses. 

This logic could be moved to a higher level controller at MachineReconciler maybe? - https://github.com/gardener/machine-controller-manager/pull/495/commits/db647117c1930acddaf5d2e6ee2bbea0affce6ac.

prashanth26 replied:
@ggaurav10 @hardikdr do provide your inputs as well.

prashanth26 replied:
> With this solution, we should simply trigger the reconciliation if the secret is NotFound, and that should be it. :)

@hardikdr with this you mean get rid of this logic right? - https://github.com/gardener/machine-controller-manager/pull/495/files#diff-3c862eb54a8e0e161b534e0c67e5379eR545-R548. And just keep returning in other cases?

hardikdr replied:
Yes, that's true.

AxiomSamarth commented:
Test case needs to added yet.

prashanth26 commented:
/reviewed ok-to-test

amshuman-kr commented:
@AxiomSamarth Thanks for the nice PR. One small thing, could you please remove the user name and issue number from the PR name? It is best to keep the PR name describing only the fix/enhancement. There are other metadata to figure out author, issue name etc.

prashanth26 commented:
Looks like all the CI tests seem to be passing 💯 . Can you also squash all your commits into a single one now that the reviews are done?

hardikdr commented:
Great PR, thanks @AxiomSamarth ...

AxiomSamarth commented:
@amshuman-kr - Thanks, Amshu! Sure, I have changed the PR name and I will follow the suggested convention.

@prashanth26 - Thanks, Prashanth. I have squashed the commits into single one.

prashanth26 commented:
PR merging is pending awaiting approval from @ggaurav10
